File: wini source/lexer.py
# lexer.py
import re
from tokens import TOKENS
from exepciones import SintaxisError

# Precompilar todos los patrones una sola vez (en vez de que re.match()
# los compile/recupere de caché en cada intento) y usarlos con match(str, pos)
# más abajo, para no tener que trocear el string en cada intento fallido.
_TOKENS_COMPILADOS = [(tipo, re.compile(patron)) for tipo, patron in TOKENS]


def _extraer_cadenas_triples(code):
    """Busca cadenas delimitadas por comillas triples (\"\"\" o ''') -incluyendo
    la variante interpolada c\"\"\"...\"\"\" / c'''...'''- y las reemplaza por
    marcadores de una sola línea envueltos en comillas normales. Así el resto
    del lexer (que trabaja línea por línea) no necesita ningún cambio.
    Se conserva el número de saltos de línea internos para no descuadrar la
    numeración de líneas del código que sigue."""
    contenidos = {}
    resultado = []
    i = 0
    n = len(code)
    contador = 0

    while i < n:
        es_c = code[i] == 'c' and code[i + 1:i + 4] in ('"""', "'''")
        es_simple = code[i:i + 3] in ('"""', "'''")

        if es_c or es_simple:
            prefijo = "c" if es_c else ""
            comillas = code[i + 1:i + 4] if es_c else code[i:i + 3]
            inicio_contenido = i + (4 if es_c else 3)

            cierre = code.find(comillas, inicio_contenido)
            if cierre == -1:
                linea = code.count("\n", 0, i) + 1
                raise SintaxisError(f"Cadena triple sin cerrar iniciada en línea {linea}")

            contenido = code[inicio_contenido:cierre]
            marcador = f"\x00T{contador}\x00"
            contenidos[marcador] = contenido
            contador += 1

            resultado.append(prefijo)
            resultado.append('"')
            resultado.append(marcador)
            resultado.append('"')
            resultado.append("\n" * contenido.count("\n"))  # conservar numeración de líneas

            i = cierre + 3
        else:
            resultado.append(code[i])
            i += 1

    return "".join(resultado), contenidos


def _restituir_triple(valor, mapa):
    """Sustituye el marcador de cadena triple por su contenido real dentro
    del valor crudo del token (que aún conserva sus comillas/prefijo 'c')."""
    if isinstance(valor, str) and "\x00T" in valor:
        for marcador, contenido in mapa.items():
            if marcador in valor:
                return valor.replace(marcador, contenido)
    return valor


def lexer(code):
    tokens = []
    # Normalizar saltos de línea
    code = code.replace("\r\n", "\n").replace("\r", "\n")
    # Extraer cadenas triples antes de tokenizar línea por línea
    code, cadenas_triples = _extraer_cadenas_triples(code)
    lines = code.split("\n")
    
    # Recorrer línea por línea
    for line_num, line in enumerate(lines, start=1):
        original_line = line
        # Expandir tabs a 4 espacios
        line = line.expandtabs(4)
        
        # Calcular indentación (espacios al inicio)
        stripped = line.lstrip(" ")
        indent = len(line) - len(stripped)
        
        # Ignorar líneas vacías o que solo tengan comentario
        if not stripped.strip() or stripped.lstrip().startswith("#"):
            continue
        
        # Token NUEVA_LINEA con valor = indentación
        tokens.append(("NUEVA_LINEA", indent, line_num, 0))
        
        # Tokenizar el contenido (sin indentación)
        content = stripped
        col = indent + 1  # columna donde empieza el contenido (1-based)
        pos = 0
        len_content = len(content)
        while pos < len_content:
            matched = False
            for token_type, regex in _TOKENS_COMPILADOS:
                # match(content, pos) usa un offset interno: no crea un
                # substring nuevo en cada intento como hacía content[pos:]
                m = regex.match(content, pos)
                if m:
                    text = m.group(0)
                    # Ignorar espacios y comentarios
                    if token_type not in ("ESPACIO", "COMENTARIO"):
                        tokens.append((token_type, text, line_num, col))
                    largo = len(text)
                    pos += largo
                    col += largo
                    matched = True
                    break
            if not matched:
                raise SintaxisError(f"Carácter inesperado '{content[pos]}' en línea {line_num}, columna {col}")

    if cadenas_triples:
        tokens = [
            (ttype, _restituir_triple(tvalue, cadenas_triples), tline, tcol)
            for (ttype, tvalue, tline, tcol) in tokens
        ]

    return tokens

File: gowini/wini/lexer.go
package wini

import (
	"strings"
)

// Lexer implementa el analizador léxico
type Lexer struct {
	tokensCompilados []DefinicionToken
}

// NewLexer crea un nuevo lexer
func NewLexer() *Lexer {
	return &Lexer{
		tokensCompilados: compilarTokens(),
	}
}

// Tokenize convierte un código fuente en una lista de tokens
func (l *Lexer) Tokenize(code string) []Token {
	// Normalizar saltos de línea
	code = strings.ReplaceAll(code, "\r\n", "\n")
	code = strings.ReplaceAll(code, "\r", "\n")

	// Extraer cadenas triples
	code, cadenasTriples := l.extraerCadenasTriples(code)

	lines := strings.Split(code, "\n")
	var tokens []Token

	// Recorrer línea por línea
	for lineNum, line := range lines {
		originalLine := line
		_ = originalLine // Usado para preservar la línea original si es necesario

		// Expandir tabs a 4 espacios
		line = strings.ReplaceAll(line, "\t", "    ")

		// Calcular indentación
		stripped := strings.TrimLeft(line, " ")
		indent := len(line) - len(stripped)

		// Ignorar líneas vacías o que solo tengan comentario
		if strings.TrimSpace(stripped) == "" || strings.HasPrefix(strings.TrimLeft(stripped, " "), "#") {
			continue
		}

		// Token NUEVA_LINEA con valor = indentación
		tokens = append(tokens, Token{
			Tipo:    NUEVA_LINEA,
			Valor:   string(rune(indent)),
			Linea:   lineNum + 1,
			Columna: 0,
		})

		// Tokenizar el contenido (sin indentación)
		content := stripped
		col := indent + 1
		pos := 0
		lenContent := len(content)

		for pos < lenContent {
			matched := false
			for _, tokenDef := range l.tokensCompilados {
				m := tokenDef.Patron.FindStringIndex(content[pos:])
				if m != nil && m[0] == 0 {
					text := content[pos : pos+m[1]]

					// Ignorar espacios y comentarios
					if tokenDef.Tipo != ESPACIO && tokenDef.Tipo != COMENTARIO {
						tokens = append(tokens, Token{
							Tipo:    tokenDef.Tipo,
							Valor:   text,
							Linea:   lineNum + 1,
							Columna: col,
						})
					}

					pos += m[1]
					col += m[1]
					matched = true
					break
				}
			}
			if !matched {
				// Error de sintaxis
				panic(&SintaxisError{
					ErrorConLinea: ErrorConLinea{
						Mensaje: "Carácter inesperado en línea " + string(rune(lineNum+1)) + ", columna " + string(rune(col)),
						Linea:   &[]int{lineNum + 1}[0],
						Columna: &[]int{col}[0],
					},
				})
			}
		}
	}

	// Restituir cadenas triples
	if len(cadenasTriples) > 0 {
		for i, token := range tokens {
			if token.Tipo == CADENA_TEXTO || token.Tipo == CADENA_INTERPOLADA {
				tokens[i].Valor = l.restituirTriple(token.Valor, cadenasTriples)
			}
		}
	}

	return tokens
}

// extraerCadenasTriples busca cadenas delimitadas por comillas triples
func (l *Lexer) extraerCadenasTriples(code string) (string, map[string]string) {
	contenidos := make(map[string]string)
	var resultado strings.Builder
	i := 0
	n := len(code)
	contador := 0

	for i < n {
		esC := i < n-3 && code[i] == 'c' && (code[i+1:i+4] == `"""` || code[i+1:i+4] == "'''")
		esSimple := i < n-2 && (code[i:i+3] == `"""` || code[i:i+3] == "'''")

		if esC || esSimple {
			prefijo := ""
			if esC {
				prefijo = "c"
			}
			comillas := ""
			if esC {
				comillas = code[i+1 : i+4]
			} else {
				comillas = code[i : i+3]
			}
			inicioContenido := i
			if esC {
				inicioContenido = i + 4
			} else {
				inicioContenido = i + 3
			}

			cierre := strings.Index(code[inicioContenido:], comillas)
			if cierre == -1 {
				linea := strings.Count(code[:i], "\n") + 1
				panic(&SintaxisError{
					ErrorConLinea: ErrorConLinea{
						Mensaje: "Cadena triple sin cerrar iniciada en línea " + string(rune(linea)),
						Linea:   &linea,
					},
				})
			}

			contenido := code[inicioContenido : inicioContenido+cierre]
			marcador := "\x00T" + string(rune(contador)) + "\x00"
			contenidos[marcador] = contenido
			contador++

			resultado.WriteString(prefijo)
			resultado.WriteString(`"`)
			resultado.WriteString(marcador)
			resultado.WriteString(`"`)
			resultado.WriteString(strings.Repeat("\n", strings.Count(contenido, "\n")))

			i = inicioContenido + cierre + 3
		} else {
			resultado.WriteByte(code[i])
			i++
		}
	}

	return resultado.String(), contenidos
}

// restituirTriple reemplaza el marcador de cadena triple por su contenido real
func (l *Lexer) restituirTriple(valor string, mapa map[string]string) string {
	if strings.Contains(valor, "\x00T") {
		for marcador, contenido := range mapa {
			if strings.Contains(valor, marcador) {
				return strings.ReplaceAll(valor, marcador, contenido)
			}
		}
	}
	return valor
}
